### ISSUE TITLE
Reading from raw bytes?

### ISSUE THREAD

**Issue by calebwin** [JuliaIO/Parquet.jl#145]
I'm downloading a Parquet file over network using AWSS3.jl. Can I parse this into a a DataFrame using Parquet.jl?

**Comment by quinnj** [comment]
@tanmaykm, I think this would be helpful to have; in [Arrow.jl](https://github.com/JuliaData/Arrow.jl/blob/b610770a75dcd03849fb3bcd4254df17e9a76153/src/table.jl#L55-#L60) and CSV.jl, we ultimately always do all the file parsing/processing on a `Vector{UInt8}` which makes it really convenient for cases like OP mentioned.

**Comment by calebwin** [comment]
@tanmaykm I don't have first-hand knowledge on this but it seems like you could easily incur overhead from reading over the network for each fetch of bytes. And so a `File` abstraction wouldn't be desirable?

**Comment by calebwin** [comment]
@tanmaykm Do you think `S3Path` from AWSS3.jl in conjunction with FilePathsBase.jl would allow for the dataset loading to work from S3 with no changes? Then the only hard part would be introducing an `AbstractFile` of some sorts.

**Comment by calebwin** [comment]
@tanmaykm @quinnj I unfortunately don't have the time to develop this at the moment - do you think this might be a valid case to just use S3FS via FUSE?

**Comment by tanmaykm** [comment]
Yes, I think S3FS via FUSE may work well in this case.

**Comment by calebwin** [comment]
@tanmaykm Okay, my only concern is - do you know if S3FS will download files to disk if it isn't using cache? I would hope that it would just download ranges of bytes to memory....

**Comment by tanmaykm** [comment]
It does seem like that from s3fs document, and I was not able to see files being written when I tried it. But it claims that using cache may make it faster and it has an option to limit cache size.

**Comment by cwiese** [comment]
I am having same issue: reading a Parquet file on S3 and hoping to benefit from reading a specific column only.  I would think this is a very popular use case.

**Comment by calebwin** [comment]
@cwiese I haven't yet tested it out for my use-case but if you read from an S3FS-mounted bucket then I believe it should only read the appropriate byte range for the column. It's not ideal since you have to set up S3FS but it could be a work-around.

Also see my comment [above](https://github.com/JuliaIO/Parquet.jl/issues/145#issuecomment-811301535). I've been looking at [DataSets.jl](https://github.com/JuliaComputing/DataSets.jl/) which looks promising but I really feel like there is a strong need for something like fsspec. S3FS looks like the best temporary solution for now.

**Comment by cwiese** [comment]
root = "s3://$(bucket)/$(path)/$(run_date)"
fxpath = "$(root)/fx_states.parquet"
p = S3Path(fxpath, config=config)
f = read(p)
using Arrow
ar = Arrow.Stream(f, pos=1)

this gets me "ERROR: ArgumentError: no arrow ipc messages found in provided input"

I figured straight to Arrow and I can create a dataframe (if needed).

**Comment by calebwin** [comment]
That definitely won't work - `Arrow.Stream` is expecting Arrow data. But you provided it Parquet data. Arrow and Parquet are different formats. And note that the format of Arrow data is same (or almost same) regardless of whether it is disk or network or in-memory.

**Comment by cwiese** [comment]
right!   But I do not see a way to construct a Parquet.File with s3Path.  If I want to read Parquest Files on AWS S3 - I will need to use Python for now - it seems.   Now the challenge is how to avoid copying all this data multiple time getting into Julia DF.

**Comment by calebwin** [comment]
> I do not see a way to construct a Parquet.File with s3Path

You can't right now. But I think what @tanmaykm is saying is that if you mount an S3 bucket to your filesystem using S3FS then you can take the location of your file in the mounted S3 bucket (which will be a normal `Path` not `S3Path`) then you can pass that into and construct a `Parquet.File`.

**Comment by layne-sadler** [comment]
"path... or file-like objects"

> pyarrow.parquet.read_table
> https://arrow.apache.org/docs/python/generated/pyarrow.parquet.read_table.html

which enables

> pandas.read_parquet()
> https://pandas.pydata.org/docs/reference/api/pandas.read_parquet.html

**Comment by jkcoxson** [comment]
I also need to read from raw bytes for a library I'm writing. What would it take to implement this?
